kivakit-core/src/main/java/com/telenav/kivakit/core/thread/StateWatcher.java

```diff
@@ -137,37 +140,40 @@ public WakeState waitFor(Predicate<State> predicate)
      * satisfied.
      */
     public WakeState waitFor(Predicate<State> predicate,
-                             LengthOfTime maximumWaitTime)
+                             Duration maximumWaitTime)
     {
+        var started = Time.now();
         return whileLocked(() ->
```

Comment:
I think you want to reverse the `whileLocked` and `while(true)` clauses.  Otherwise this is going to block until the lock is acquired, regardless of what was passed in for a timeout.

And/or you want a version of `whileLocked()` that takes a timeout.

Actually, it's going to block until it gets the lock no matter what order those clauses are in.

Probably it should not acquire the lock (in which case the `volatile` above _is_ needed) - but acquiring the lock in a loop might be cleaner, as long as you can provide a timeout for how long you want to wait for it.

Reply:
Well, removing the lock makes the most sense, but it's protecting access to *waiters*. What do you think about making that collection synchronized and removing the lock? (the access in signal() would have to make a defensive copy).

Reply:
I removed the lock and synced the collection.
Have a look. Should be resolved now.